README.md
# CSE210-Template (CHANGE ME!!!)

BYUI CSE-210 Template Repo (CHANGE ME!!!)

## How to use this project

1. Fork the repository (using the GitHub 'fork' button on the repo page)

2. Switch to the 'dev' branch. If you don't know how, Google is your friend [here](https://bfy.tw/SPc6)

3. Contribute your code.

4. Once you have finished your code contribution, make a commit (or multiple commits), see [here](https://bfy.tw/SPc5) for how to make a commit.

5. Submit a pull request, instructions on how to do so [here](https://bfy.tw/SPc7).

6. Repeat.

7. Profit?

## Changelogs

* To see the changelogs, and to add to it, please modify this file: [RELEASE-NOTES](RELEASE-NOTES.md)

* If you don't know how to use MarkDown (GitHub's text editor language), please see here for a reference guide: [link](https://www.markdownguide.org/)


* To add to the group ToDo list, please create an "Issue," with the "Issues" tab of the GitHub repository. Please choose the correct category and all should go well.
  * If you are requesting a feature, please assign yourself to the issue. This will be how we track our "TODO's."
  * If you are submiting a bug report, please give **AS MUCH INFORMATION AS POSSIBLE**, and assign yourself to the issue. Others who can handle it will assign themselves when they can tackle it.
  * If you have a question (and Google hasn't produced any good results), please create an issue with the category of Question. Again, the more detail, the better. Also, assign yourself to this, again.


RELEASE-NOTES.md
# RELEASE NOTES

## 0.0.1

* Initial release


Greed Game/__main__.py
import os
import random
from tkinter.tix import CELL

from game.casting.actor import Actor
from game.casting.cast import Cast
from game.casting.gem import Gem
from game.casting.rock import Rock

from game.directing.director import Director

from game.services.keyboard_service import KeyboardService
from game.services.video_service import VideoService

from game.shared.color import Color
from game.shared.point import Point


FRAME_RATE = 60
MAX_X = 900
MAX_Y = 600
CELL_SIZE = 15
FONT_SIZE = 30
COLS = MAX_X / CELL_SIZE
ROWS = MAX_Y / CELL_SIZE
CAPTION = "Rocks and Gems game"
DATA_PATH = os.path.dirname(os.path.abspath(__file__)) + "/data/messages.txt"
WHITE = Color(255, 255, 255)
DEFAULT_ARTIFACTS = 10
START_ROCKS = 2
START_GEMS = 5

def main():
    
    # create the cast
    cast = Cast()
    
    # create the banner
    banner = Actor()
    banner.set_text("")
    banner.set_font_size(FONT_SIZE)
    banner.set_color(WHITE)
    banner.set_position(Point(CELL_SIZE, 0))
    cast.add_actor("banners", banner)
    
    # create the robot
    x = int(MAX_X / 2)
    y = MAX_Y - FONT_SIZE
    position = Point(x, y)

    robot = Actor()
    robot.set_text("#")
    robot.set_font_size(FONT_SIZE)
    robot.set_color(WHITE)
    robot.set_position(position)
    cast.add_actor("robots", robot)

    for i in range(START_ROCKS):
        rock = Rock()
        cast.add_actor("rocks", rock)
    
    for i in range(START_GEMS):
        gem = Gem()
        cast.add_actor("gems", gem)

    # start the game
    keyboard_service = KeyboardService(CELL_SIZE)
    video_service = VideoService(CAPTION, MAX_X, MAX_Y, CELL_SIZE, FRAME_RATE)
    director = Director(keyboard_service, video_service)
    director.start_game(cast)

if __name__ == "__main__":
    main()

Greed Game/game/casting/actor.py
from game.shared.color import Color
from game.shared.point import Point


class Actor:
    """A visible, moveable thing that participates in the game. 
    
    The responsibility of Actor is to keep track of its appearance, position and velocity in 2d 
    space.

    Attributes:
        _text (string): The text to display
        _font_size (int): The font size to use.
        _color (Color): The color of the text.
        _position (Point): The screen coordinates.
        _velocity (Point): The speed and direction.
    """

    def __init__(self):
        """Constructs a new Actor."""
        self._text = ""
        self._point_value = 0
        self._font_size = 20
        self._color = Color(255, 255, 255)
        self._position = Point(0, 0)
        self._velocity = Point(0, 0)

    def get_color(self):
        """Gets the actor's color as a tuple of three ints (r, g, b).
        
        Returns:
            Color: The actor's text color.
        """
        return self._color

    def get_font_size(self):
        """Gets the actor's font size.
        
        Returns:
            Point: The actor's font size.
        """
        return self._font_size

    def get_position(self):
        """Gets the actor's position in 2d space.
        
        Returns:
            Point: The actor's position in 2d space.
        """
        return self._position
    
    def get_text(self):
        """Gets the actor's textual representation.
        
        Returns:
            string: The actor's textual representation.
        """
        return self._text

    def get_velocity(self):
        """Gets the actor's speed and direction.
        
        Returns:
            Point: The actor's speed and direction.
        """
        return self._velocity
    
    def move_next(self, max_x, max_y):
        """Moves the actor to its next position according to its velocity. Will wrap the position 
        from one side of the screen to the other when it reaches the given maximum x and y values.
        
        Args:
            max_x (int): The maximum x value.
            max_y (int): The maximum y value.
        """

        x = (self._position.get_x() + self._velocity.get_x()) % max_x
        y = (self._position.get_y() + self._velocity.get_y()) % max_y
        self._position = Point(x, y)

    def set_color(self, color):
        """Updates the color to the given one.
        
        Args:
            color (Color): The given color.
        """
        self._color = color

    def set_position(self, position):
        """Updates the position to the given one.
        
        Args:
            position (Point): The given position.
        """
        self._position = position
    
    def set_font_size(self, font_size):
        """Updates the font size to the given one.
        
        Args:
            font_size (int): The given font size.
        """
        self._font_size = font_size
    
    def set_text(self, text):
        """Updates the text to the given value.
        
        Args:
            text (string): The given value.
        """
        self._text = text

    def set_velocity(self, velocity):
        """Updates the velocity to the given one.
        
        Args:
            velocity (Point): The given velocity.
        """
        self._velocity = velocity

Greed Game/game/casting/object.py
from game.casting.actor import Actor
import random
from game.shared.point import Point
from game.shared.color import Color

VERT_SPEED = 5
HOR_SPEED = 1

class Object(Actor):
    """
    An item of cultural or historical interest. 
    
    The responsibility of an Artifact is to provide a message about itself.

    Attributes:
        _message (string): A short description about the artifact.
    """
    def __init__(self):
        super().__init__()
        x = random.randint(1, 60)
        y = random.randint(1, 40)
        position = Point(x, y)
        position = position.scale(15)

        r = random.randint(0, 255)
        g = random.randint(0, 255)
        b = random.randint(0, 255)
        color = Color(r, g, b)
    
        self._font_size = 20
        self._color = color
        self._position = position
        
        dy = random.randint(int(VERT_SPEED / 3), VERT_SPEED)
        dx = random.randint(- HOR_SPEED, HOR_SPEED)
        self._velocity = Point(dx, dy)

Greed Game/game/services/video_service.py
import pyray


class VideoService:
    """Outputs the game state. The responsibility of the class of objects is to draw the game state 
    on the screen. 
    """

    def __init__(self, caption, width, height, cell_size, frame_rate, debug = False):
        """Constructs a new VideoService using the specified debug mode.
        
        Args:
            debug (bool): whether or not to draw in debug mode.
        """
        self._caption = caption
        self._width = width
        self._height = height
        self._cell_size = cell_size
        self._frame_rate = frame_rate
        self._debug = debug

    def close_window(self):
        """Closes the window and releases all computing resources."""
        pyray.close_window()

    def clear_buffer(self):
        """Clears the buffer in preparation for the next rendering. This method should be called at
        the beginning of the game's output phase.
        """
        pyray.begin_drawing()
        pyray.clear_background(pyray.BLACK)
        if self._debug == True:
            self._draw_grid()
    
    def draw_actor(self, actor):
        """Draws the given actor's text on the screen.

        Args:
            actor (Actor): The actor to draw.
        """ 
        text = actor.get_text()
        x = actor.get_position().get_x()
        y = actor.get_position().get_y()
        font_size = actor.get_font_size()
        color = actor.get_color().to_tuple()
        pyray.draw_text(text, x, y, font_size, color)
        
    def draw_actors(self, actors):
        """Draws the text for the given list of actors on the screen.

        Args:
            actors (list): A list of actors to draw.
        """ 
        for actor in actors:
            self.draw_actor(actor)
    
    def flush_buffer(self):
        """Copies the buffer contents to the screen. This method should be called at the end of
        the game's output phase.
        """ 
        pyray.end_drawing()

    def get_cell_size(self):
        """Gets the video screen's cell size.
        
        Returns:
            Grid: The video screen's cell size.
        """
        return self._cell_size

    def get_height(self):
        """Gets the video screen's height.
        
        Returns:
            Grid: The video screen's height.
        """
        return self._height

    def get_width(self):
        """Gets the video screen's width.
        
        Returns:
            Grid: The video screen's width.
        """
        return self._width

    def is_window_open(self):
        """Whether or not the window was closed by the user.

        Returns:
            bool: True if the window is closing; false if otherwise.
        """
        return not pyray.window_should_close()

    def open_window(self):
        """Opens a new window with the provided title.

        Args:
            title (string): The title of the window.
        """
        pyray.init_window(self._width, self._height, self._caption)
        pyray.set_target_fps(self._frame_rate)



Greed Game/game/services/keyboard_service.py
import pyray
from game.shared.point import Point


class KeyboardService:
    """Detects player input. 
    
    The responsibility of a KeyboardService is to detect player key presses and translate them into 
    a point representing a direction.

    Attributes:
        cell_size (int): For scaling directional input to a grid.
    """

    def __init__(self, cell_size = 1):
        """Constructs a new KeyboardService using the specified cell size.
        
        Args:
            cell_size (int): The size of a cell in the display grid.
        """
        self._cell_size = cell_size

    def get_direction(self):
        """Gets the selected direction based on the currently pressed keys.

        Returns:
            Point: The selected direction.
        """
        dx = 0
        dy = 0

        if pyray.is_key_down(pyray.KEY_LEFT):
            dx = -1
        
        if pyray.is_key_down(pyray.KEY_RIGHT):
            dx = 1
        
        if pyray.is_key_down(pyray.KEY_UP):
            dy = 0
        
        if pyray.is_key_down(pyray.KEY_DOWN):
            dy = 0

        direction = Point(dx, dy)
        direction = direction.scale(self._cell_size)
        
        return direction

Greed Game/game/directing/director.py
import math
from game.casting.gem import Gem
from game.casting.rock import Rock
from game.shared.point import Point
import random

HIT_RADIUS = 50

class Director:
    """A person who directs the game. 
    
    The responsibility of a Director is to control the sequence of play.

    Attributes:
        _keyboard_service (KeyboardService): For getting directional input.
        _video_service (VideoService): For providing video output.
    """

    def __init__(self, keyboard_service, video_service):
        """Constructs a new Director using the specified keyboard and video services.
        
        Args:
            keyboard_service (KeyboardService): An instance of KeyboardService.
            video_service (VideoService): An instance of VideoService.
        """
        self._keyboard_service = keyboard_service
        self._video_service = video_service
        self._score = 0
        
    def start_game(self, cast):
        """Starts the game using the given cast. Runs the main game loop.

        Args:
            cast (Cast): The cast of actors.
        """
        self._video_service.open_window()
        while self._video_service.is_window_open():
            self._get_inputs(cast)
            self._do_updates(cast)
            self._do_outputs(cast)
        self._video_service.close_window()

    def _get_inputs(self, cast):
        """Gets directional input from the keyboard and applies it to the robot.
        
        Args:
            cast (Cast): The cast of actors.
        """
        robot = cast.get_first_actor("robots")
        velocity = self._keyboard_service.get_direction()
        robot.set_velocity(velocity)

    def _do_updates(self, cast):
        """Updates the robot's position and resolves any collisions with artifacts.
        
        Args:
            cast (Cast): The cast of actors.
        """
        banner = cast.get_first_actor("banners")
        robot = cast.get_first_actor("robots")
        gems = cast.get_actors("gems")
        rocks = cast.get_actors("rocks")
        current_position = robot.get_position()
        cur_x = current_position.get_x()
        cur_y = current_position.get_y()

        banner.set_text(f"Score: {self._score}")
        max_x = self._video_service.get_width()
        max_y = self._video_service.get_height()
        robot.move_next(max_x, max_y)
        
        for gem in gems:
            gem_position = gem.get_position()
            cur_art_x = gem_position.get_x()
            cur_art_y = gem_position.get_y()
            gem.move_next(900, 600)

            if (math.sqrt(((cur_art_x-cur_x)**2) + ((cur_art_y-cur_y)**2))) < HIT_RADIUS:
                self._score += gem._point_value  
                cast.remove_actor("gems", gem)
                if int(len(cast.get_actors("gems"))) < 12:
                    gem = Gem()
                    dy = random.randint(3+ int(self._score/30), 15)
                    dx = random.randint(- 1, 1)
                    velocity = Point(dx, dy)
                    gem.set_velocity(velocity)
                    cast.add_actor("gems", gem)
                    gem = Gem()
                    dy = random.randint(3+ int(self._score/30), 15)
                    dx = random.randint(- 1, 1)
                    velocity = Point(dx, dy)
                    gem.set_velocity(velocity)
                    cast.add_actor("gems", gem)

        for rock in rocks:
            rock_position = rock.get_position()
            cur_art_x = rock_position.get_x()
            cur_art_y = rock_position.get_y()
            rock.move_next(900, 600)
            if self._score < 50:
                dy = int(rock._velocity.get_y())
                if dy < 6:
                    rock._velocity._y = int(dy * (1 + self._score / 10))
            
            if (math.sqrt(((cur_art_x-cur_x)**2) + ((cur_art_y-cur_y)**2))) < HIT_RADIUS:
                self._score += rock._point_value
                cast.remove_actor("rocks", rock)
                if int(len(cast.get_actors("rocks"))) < 10:
                    rock = Rock()
                    dy = random.randint(3+ int(self._score/30), 15)
                    dx = random.randint(- 1, 1)
                    velocity = Point(dx, dy)
                    rock.set_velocity(velocity)
                    cast.add_actor("rocks", rock)
                    rock = Rock()
                    dy = random.randint(3+ int(self._score/30), 15)
                    dx = random.randint(- 1, 1)
                    velocity = Point(dx, dy)
                    rock.set_velocity(velocity)
                    cast.add_actor("rocks", rock)

    def _do_outputs(self, cast):
        """Draws the actors on the screen.
        
        Args:
            cast (Cast): The cast of actors.
        """
        self._video_service.clear_buffer()
        actors = cast.get_all_actors()
        self._video_service.draw_actors(actors)
        self._video_service.flush_buffer()